[DRAFT] Add per shard toppartitions api

igorribeiroduarte opened the pull request.

mykaul commented:
CC @amnonh

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3105/

amnonh commented:
@igorribeiroduarte Can you add a description for what you did and what you want to achieve?

igorribeiroduarte commented:
> @igorribeiroduarte Can you add a description for what you did and what you want to achieve?

Done

amnonh commented:
Some general notes, I'm not sure about the per-shard implementation, if the idea is to add the shard id to the result, than it can be on by default (just need to make sure nothing is broken with JMX (i don't think it will).

If it's for debuging specifc shard (I had that request last time more functionality was added) then it the shard id flag, should hold the specific shard and only take the infromation from it (when you are trying to understand what a specific shard is doing).

In real scenario, we execpt that the top-k will be at least in the hundreds (if not way more).
To lower the performace issuse we can use play with the probability if to concider a partition or not (e.g. sample only one in 20, or one in hundred) This will not return the actuall number, but it will return with good probebility the topk

There are some majic numbers in the code, I suggest move them to configuration.

igorribeiroduarte commented:
@amnonh 

> Some general notes, I'm not sure about the per-shard implementation, if the idea is to add the shard id to the result, than it can be on by default (just need to make sure nothing is broken with JMX (i don't think it will).
> If it's for debuging specifc shard (I had that request last time more functionality was added) then it the shard id flag, should hold the specific shard and only take the infromation from it (when you are trying to understand what a specific shard is doing).

The idea is to do both. The shard id was permanently added to the results and it's not optional.
What is optional is to get the results per shard instead of per cluster (as it is done today). So, if the `per_shard` flag is set to `true`, `list_size * smp::count` results will be returned instead of just `list_size`.

> In real scenario, we execpt that the top-k will be at least in the hundreds (if not way more).
To lower the performace issuse we can use play with the probability if to concider a partition or not (e.g. sample only one in 20, or one in hundred) This will not return the actuall number, but it will return with good probebility the topk

Indeed! Thanks for this suggestion.
I added this to the patch series. Could you please re-review?

> There are some majic numbers in the code, I suggest move them to configuration.

Done

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3448/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3452/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3454/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3455/

denesb commented on db/data_listeners.hh:
```diff
@@ -45,6 +54,16 @@ public:
+
+    virtual bool sample_next_query() {
+        return _normalized_sampling_probability != 0 && _gen() < _normalized_sampling_probability;
+    }
```
Why are these virtual?

igorribeiroduarte replied:
I thought it would make sense to make them virtual in order to allow other derived data_listeners that we might create in the future to override these methods.

denesb replied:
I think you can drop the virtual. Those future users (if they materialize), can re-add it themselves. Until then, no need to pay the cost of a virtual call (however minuscule that might be).

igorribeiroduarte replied:
Done

denesb commented on db/config.cc:
```diff
@@ -1024,6 +1024,10 @@ db::config::config(std::shared_ptr<db::extensions> exts)
         "Regular expression used by CertificateAuthenticator to extract role name from an accepted transport authentication certificate subject info.")
+    , persistent_toppartitions_publish_interval_sec(this, "persistent_toppartitions_publish_interval", liveness::LiveUpdate, value_status::Used, 10, "Controls how often the toppartitions results will be published to the logs.")
+    , persistent_toppartitions_capacity(this, "persistent_toppartitions_capacity", liveness::LiveUpdate, value_status::Used, 256, "Controls the per shard capacity that will be used to calculate the toppartitions.")
+    , persistent_toppartitions_list_size(this, "persistent_toppartitions_list_size", liveness::LiveUpdate, value_status::Used, 20, "Controls how many partitions per shard will be showed in the toppartitions results.")
```
Not clear what the difference is between list size and capacity. I though they are the same but apparently not. What is "capacity"?

igorribeiroduarte replied:
The `list_size` is the number of partitions that we're going to show in the results.
The `capacity` represents the maximum number of elements that we're going to track while we're calculating the toppartitions. We do this to limit the amount of space that is going to be used while sampling the toppartitions. The general idea is explained here -> https://github.com/scylladb/scylladb/issues/2811#issuecomment-420212450

denesb commented on replica/database.hh:
```diff
@@ -1432,6 +1433,7 @@ private:
 
     friend db::data_listeners;
     std::unique_ptr<db::data_listeners> _data_listeners;
+    std::unique_ptr<db::toppartitions_data_listener> _tp_listener;
```
Why not use the member already available above?

igorribeiroduarte replied:
Because the member above of the type `data_listeners` is an abstraction of a `set` of a generic `data_listener` and what I need here is a `toppartitions_data_listener` in order to persistently sample the toppartitions.

denesb commented on replica/database.cc:
```diff
@@ -393,6 +395,24 @@ database::database(const db::config& cfg, database_config dbcfg, service::migrat
+    _tp_timer.arm_periodic(std::chrono::seconds(_cfg.persistent_toppartitions_publish_interval_sec()));
+}
+
+void database::on_tp_timer() {
```
Please avoid abbreviations like this in a general class like database.

igorribeiroduarte replied:
Fixed.

denesb commented on replica/database.cc:
```diff
@@ -393,6 +395,24 @@ database::database(const db::config& cfg, database_config dbcfg, service::migrat
+
+    for (auto& d: _tp_listener->get_top_k_read().top(list_size).values) {
+        auto partition = ("(" + d.item.schema->ks_name() + ":" + d.item.schema->cf_name() + ") ") + sstring(d.item);
+        metricslog.debug("Toppartitions - Read - Partition [{}]: {}", partition, d.count);
```
If we are doing all this work of gathering the metrics and generating a report, we might as well use `info` so it is actually appears in the logs.

igorribeiroduarte replied:
Done.

denesb commented on replica/database.hh:
```diff
@@ -1434,6 +1434,8 @@ private:
     friend db::data_listeners;
     std::unique_ptr<db::data_listeners> _data_listeners;
     std::unique_ptr<db::toppartitions_data_listener> _tp_listener;
+    timer<seastar::lowres_clock> _tp_timer;
```
Pleas avoid using abbreviations in the database class.

igorribeiroduarte replied:
Fixed.

denesb commented on replica/database.cc:
```diff
@@ -415,6 +420,18 @@ void database::on_tp_timer() {
+    dblog.info("Updating toppartitions listener");
+
+    if (!_tp_listener->set_sampling_probability(_cfg.persistent_toppartitions_sampling_probability())) {
+        dblog.error("Failed to apply persistent toppartitions changes. Sampling probability must be in a [0,1] range");
```
Please also log the sampling probability, so we know what the bad value is.

igorribeiroduarte replied:
Done.

denesb commented:
We will also need to do a quick benchmark with `scylla perf-simple-query` to see how much this affects performance when turned off and turned on.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3561/

igorribeiroduarte commented:
v2:
* Removed abbreviations
* Used `info` level to print the metrics instead of `debug`
* Added the sampling probability to the error message in the `update_toppartitions_listener` method

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3562/

igorribeiroduarte commented:
> We will also need to do a quick benchmark with `scylla perf-simple-query` to see how much this affects performance when turned off and turned on.

Command used: `scylla perf-simple-query --duration 15 -m 1G`

Turned off:
```
Creating 10000 partitions...
818120.59 tps ( 64.0 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
822022.12 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
821354.57 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
822927.37 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
826640.86 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
821121.29 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
823002.27 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
820346.14 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
817703.33 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
822741.75 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
815793.24 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
825307.39 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
818746.70 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
823137.64 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
827700.06 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)

median 822022.12 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
median absolute deviation: 1675.97
maximum: 827700.06
minimum: 815793.24
```

Turned on - 100%:
```
Creating 10000 partitions...
765237.83 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
764504.27 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
766843.36 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
769300.89 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
762642.68 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
765109.32 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
767018.54 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
767952.21 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
761235.14 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
767780.30 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
765645.71 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
757987.80 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
765418.79 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
766479.64 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
764074.87 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)

median 765418.79 tps ( 70.6 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
median absolute deviation: 1424.56
maximum: 769300.89
minimum: 757987.80
```

Turned on - 50%:
```
Creating 10000 partitions...
783651.97 tps ( 67.3 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
794546.53 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
791325.80 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
786492.33 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
792217.70 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
787785.52 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
783920.85 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
798933.87 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
787547.92 tps ( 67.3 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
783871.93 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
793314.06 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
788928.40 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
789356.63 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
791937.40 tps ( 67.3 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
788170.42 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)

median 788928.40 tps ( 67.2 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
median absolute deviation: 3008.99
maximum: 798933.87
minimum: 783651.97
```

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3575/

mykaul commented on db/config.cc:
```diff
@@ -1026,6 +1026,10 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , minimum_replication_factor_warn_threshold(this, "minimum_replication_factor_warn_threshold", liveness::LiveUpdate, value_status::Used,  3, "")
     , maximum_replication_factor_warn_threshold(this, "maximum_replication_factor_warn_threshold", liveness::LiveUpdate, value_status::Used, -1, "")
     , maximum_replication_factor_fail_threshold(this, "maximum_replication_factor_fail_threshold", liveness::LiveUpdate, value_status::Used, -1, "")
+    , persistent_toppartitions_publish_interval_sec(this, "persistent_toppartitions_publish_interval", liveness::LiveUpdate, value_status::Used, 10, "Controls how often the toppartitions results will be published to the logs.")
```
in ms? seconds? minutes?

igorribeiroduarte replied:
In seconds.
I updated the description.

mykaul commented on db/config.cc:
```diff
@@ -1026,6 +1026,10 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , maximum_replication_factor_fail_threshold(this, "maximum_replication_factor_fail_threshold", liveness::LiveUpdate, value_status::Used, -1, "")
+    , persistent_toppartitions_publish_interval_sec(this, "persistent_toppartitions_publish_interval", liveness::LiveUpdate, value_status::Used, 10, "Controls how often the toppartitions results will be published to the logs.")
+    , persistent_toppartitions_capacity(this, "persistent_toppartitions_capacity", liveness::LiveUpdate, value_status::Used, 256, "Controls the per shard capacity that will be used to calculate the toppartitions.")
+    , persistent_toppartitions_list_size(this, "persistent_toppartitions_list_size", liveness::LiveUpdate, value_status::Used, 20, "Controls how many partitions per shard will be showed in the toppartitions results.")
```
```suggestion
    , persistent_toppartitions_list_size(this, "persistent_toppartitions_list_size", liveness::LiveUpdate, value_status::Used, 20, "Controls how many partitions per shard will be shown in the toppartitions results.")
```

igorribeiroduarte replied:
Fixed

igorribeiroduarte commented:
v4:
* Updated the descriptions of the config values in `db/config.cc` with the suggestions from @mykaul 
* Fixed a small compilation error in `top_k_test.cc`
* Squashed the commit adding the `sampling_probability` to `table_listener` in `data_listeners_test` and the commit which changed the `data_listener` constructor

igorribeiroduarte commented:
> The performance impact is prohibitive.

@avikivity , the performance impact is big when setting the probability to 100%, but since this is configurable, we could sample a much smaller value, e.g.: 10%.
Also, the feature is disabled by default.

mykaul commented:
@igorribeiroduarte do you have numbers before/after this patch (disabled)?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3580/

igorribeiroduarte commented:
> @igorribeiroduarte do you have numbers before/after this patch (disabled)?

@mykaul , yes:

without my patches (0656810c287788fb5a7ae763b79ebe96f0f38114):
```
Creating 10000 partitions...
835305.47 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
834859.68 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
837371.31 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
831991.65 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
831963.08 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
825155.44 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
830203.04 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
838528.19 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
836528.86 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
840696.91 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
834549.91 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
813795.75 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
815375.20 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
826572.19 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
836188.51 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)

median 834549.91 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
median absolute deviation: 2821.40
maximum: 840696.91
minimum: 813795.75
```

With my patches, but with the feature disabled:
```
Creating 10000 partitions...
844489.25 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
840123.29 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
839990.22 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
840563.06 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
838409.48 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
839447.04 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
839414.60 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
840627.91 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
837750.70 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
838833.88 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
838225.80 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
838348.69 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
839831.41 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
838178.72 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
838598.82 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)

median 839414.60 tps ( 63.9 allocs/op,  17.0 tasks/op,       0 insns/op,        0 errors)
median absolute deviation: 1005.13
maximum: 844489.25
minimum: 837750.70
```

avikivity commented:
> > The performance impact is prohibitive.
> 
> @avikivity , the performance impact is big when setting the probability to 100%, but since this is configurable, we could sample a much smaller value, e.g.: 10%. Also, the feature is disabled by default.

How do you operate it, then?

Note that toppartitions needs some tuning to work correctly, the default population is pretty bad. The right population depends on the workload and period.

avikivity commented:
I see you mention the configuration, but changing the configuration for something transient looks tedious, no?

avikivity commented:
Why not add a flag to nodetool? `nodetool toppartitions --output-to-log`?

igorribeiroduarte commented:
> > > The performance impact is prohibitive.
> > 
> > 
> > @avikivity , the performance impact is big when setting the probability to 100%, but since this is configurable, we could sample a much smaller value, e.g.: 10%. Also, the feature is disabled by default.
> 
> How do you operate it, then?
> 
> Note that toppartitions needs some tuning to work correctly, the default population is pretty bad. The right population depends on the workload and period.

@avikivity, Could you please explain a little more about why this tuning would be necessary here? AFAIU the use case when the results are not very reliable is when you have your partitions more or less balanced and because of this you have all partitions being evicted all the time.
However, when you have an imbalanced use case, with specific partitions receiving considerably more hits than others these partitions will appear on the top regardless of the computed count being precise or not and this is exactly the use case for which this feature is designed for.
Also note that the `top_k` values are reset every `persistent_toppartitions_publish_interval_sec` seconds, so we won't be propagating the calculated error forever.

> I see you mention the configuration, but changing the configuration for something transient looks tedious, no?

It's not clear to me why this would be transient. The idea is to have this feature enabled all the time (with a small probability set) so that it'll make easier for users to find hot partitions. Note that a lot of times when a customer asks us to investigate a performance issue this issue is not happening at the moment, so it makes sense to have this enabled all the time IMO.


> Why not add a flag to nodetool? nodetool toppartitions --output-to-log?

I think this would make sense if the goal was to have the feature enabled for a limited amount of time, but the idea is to have this enabled all the time and to have a graph in `scylla-monitoring` plotting the values, just like we do for any other metric.

igorribeiroduarte commented:
ping @avikivity @denesb @mykaul

igorribeiroduarte commented:
ping2 @avikivity @denesb @mykaul

denesb commented:
> ping2 @avikivity @denesb @mykaul

You need to convince @avikivity that the performance cost is worth paying. The argument that this will help investigate problems won't work. For this to be acceptable, you will probably need to reduce the performance impact somehow and come up with a default configuration (sample chance) that is both useful and has acceptable (very low) impact on performance.

I would start by looking into the additional allocations sampling adds to the reads. We worked very hard to reduce the allocation count and to keep it at the current level, we won't allow it to jump by 10% for a debugging feature. Maybe there is a way to pre-allocate the data-structures needed for sampling, so they don't add allocations to each request.

vladzcloudius commented:
> > ping2 @avikivity @denesb @mykaul
> 
> You need to convince @avikivity that the performance cost is worth paying. The argument that this will help investigate problems won't work. For this to be acceptable, you will probably need to reduce the performance impact somehow and come up with a default configuration (sample chance) that is both useful and has acceptable (very low) impact on performance.
> 
> I would start by looking into the additional allocations, that the sampling adds to the reads. We worked very hard to reduce the allocation count and to keep it at the current level, we won't allow it to jump by 10% for a debugging feature. Maybe there is a way to pre-allocate the data-structures needed for sampling, so they don't add allocations to each request.

@denesb @avikivity the code this PR adds is disabled by default and Igor has already shown that by default (when the feature is disabled) there is zero performance impact: https://github.com/scylladb/scylladb/pull/15065#issuecomment-1714369913

So, I believe this had addressed the "performance impact" objection. Didn't it?
Are there any other reasons not to get this merged?

denesb commented:
> @denesb @avikivity the code this PR adds is disabled by default and Igor has already shown that by default (when the feature is disabled) there is zero performance impact: [#15065 (comment)](https://github.com/scylladb/scylladb/pull/15065#issuecomment-1714369913)
> 
> So, I believe this had addressed the "performance impact" objection. Didn't it? Are there any other reasons not to get this merged?

I don't understand. You say this would be disabled, but @igorribeiroduarte wrote above:

> The idea is to have this feature enabled all the time (with a small probability set) so that it'll make easier for users to find hot partitions. Note that a lot of times when a customer asks us to investigate a performance issue this issue is not happening at the moment, so it makes sense to have this enabled all the time IMO.

To me it looks like this is very much intended to be on all the time, with a low probability (which Duarte didn't specify). So we do need to address the performance impact, with the default settings at least.

igorribeiroduarte commented:
> > @denesb @avikivity the code this PR adds is disabled by default and Igor has already shown that by default (when the feature is disabled) there is zero performance impact: [#15065 (comment)](https://github.com/scylladb/scylladb/pull/15065#issuecomment-1714369913)
> > So, I believe this had addressed the "performance impact" objection. Didn't it? Are there any other reasons not to get this merged?
> 
> I don't understand. You say this would be disabled, but @igorribeiroduarte wrote above:
> 
> > The idea is to have this feature enabled all the time (with a small probability set) so that it'll make easier for users to find hot partitions. Note that a lot of times when a customer asks us to investigate a performance issue this issue is not happening at the moment, so it makes sense to have this enabled all the time IMO.
> 
> To me it looks like this is very much intended to be on all the time, with a low probability (which Duarte didn't specify). So we do need to address the performance impact, with the default settings at least.

* @denesb, the feature is disable by default, so there will be no performance impact unless the user decides to enable it.
* **if** the user decides to enable the feature, they will be able to set a `sampling_probability` and see how much of the performance impact they are willing to accept, and then the user can choose to have this enabled all the time in the way I explained in the comment you mentioned.

denesb commented:
> * @denesb, the feature is disable by default, so there will be no performance impact unless the user decides to enable it.
> 
> * **if** the user decides to enable the feature, they will be able to set a `sampling_probability` and see how much of the performance impact they are willing to accept, and then the user can choose to have this enabled all the time in the way I explained in the comment you mentioned.

Ok, if this will be meant to be turned on on-demand, then I'm fine with a performance impact. Although I would still prefer if we worked on that so this can be on all the time. But we can do that in a follow-up. If the performance impact is too high, people will only enable it once there is trouble and that goes against the entire goal of this PR, which is to have this turned on all the time.

You still need to address Avi's comments w.r.t. accurate reporting.

denesb commented:
@igorribeiroduarte please add the motivation and the intended usage to the cover-letter.

mykaul commented:
I'm pretty sure the whole idea is that it would run regularly, not on/off, but with 1% (or even less, if that's possible) probability.

igorribeiroduarte commented:
> @igorribeiroduarte please add the motivation and the intended usage to the cover-letter.

@denesb, Done.
I see that my initial cover-letter was not very clear and wasn't focusing in the important details of the PR. Sorry for that.
I hope it's more clear now.

igorribeiroduarte commented:
> > > > > The performance impact is prohibitive.
> > > > 
> > > > 
> > > > @avikivity , the performance impact is big when setting the probability to 100%, but since this is configurable, we could sample a much smaller value, e.g.: 10%. Also, the feature is disabled by default.
> > > 
> > > 
> > > How do you operate it, then?
> > > Note that toppartitions needs some tuning to work correctly, the default population is pretty bad. The right population depends on the workload and period.
> > 
> > 
> > @avikivity, Could you please explain a little more about why this tuning would be necessary here? AFAIU the use case when the results are not very reliable is when you have your partitions more or less balanced and because of this you have all partitions being evicted all the time.
> 
> The partitions don't have to be balanced to get unreliable results, just many.
> 
> > However, when you have an imbalanced use case, with specific partitions receiving considerably more hits than others these partitions will appear on the top regardless of the computed count being precise or not and this is exactly the use case for which this feature is designed for. Also note that the `top_k` values are reset every `persistent_toppartitions_publish_interval_sec` seconds, so we won't be propagating the calculated error forever.
> 
> It can happen than unbalanced partitions and badly sized top-k still report the overloaded partitions, but it doesn't have to happen. If we report something, we want to make it reliable.
> 
> > > I see you mention the configuration, but changing the configuration for something transient looks tedious, no?
> > 
> > 
> > It's not clear to me why this would be transient. The idea is to have this feature enabled all the time (with a small probability set) so that it'll make easier for users to find hot partitions. Note that a lot of times when a customer asks us to investigate a performance issue this issue is not happening at the moment, so it makes sense to have this enabled all the time IMO.
> 
> Well only if the performance impact is negligible.
> 
> > > Why not add a flag to nodetool? nodetool toppartitions --output-to-log?
> > 
> > 
> > I think this would make sense if the goal was to have the feature enabled for a limited amount of time, but the idea is to have this enabled all the time and to have a graph in `scylla-monitoring` plotting the values, just like we do for any other metric.

@avikivity, In most cases users care less about the specific partition data precision and more about the relation between different partitions because even if we don't have the real number of hits that a partition is receiving, we're still able to identify the hot partitions.
As I mentioned in my comment to Botond, I understand that my previous cover-letter was not very clear and wasn't addressing things like the impact on performance and the usage. I've updated it and added an image showing how it'll be useful even if the number of partition hits is not accurate. Could you please take a look again?

avikivity commented:
If the top-k configuration is wrong, you'll get incorrect data. In some cases you'll see the hot partition even with incorrect configuration, but in other cases it can be lost.

Also, I think spamming the logs with this is wrong. It should be made available over a cql virtual table, and then sampled/displayed/logged via the monitoring stack.

vladzcloudius commented:
> If the top-k configuration is wrong, you'll get incorrect data. In some cases you'll see the hot partition even with incorrect configuration, but in other cases it can be lost.

Could you, please, clarify what do you refer by "incorrect configuration" here?

> 
> Also, I think spamming the logs with this is wrong. It should be made available over a cql virtual table, and then sampled/displayed/logged via the monitoring stack.

Using syslog + Locki allows us a very fast integration with the existing Scylla Monitoring stack (this is what you see on a screenshot). 

It's unclear to me how to store historical data originally read from the virtual table.
And the whole point of a Top-K graph is to have a historical data and not only the current state like we have for a `system.clients` table today.

@amnonh is there a way to push the data from the virtual table into Prometheus somehow?

Also, @avikivity, note that these new printouts go in a separate logger and hence can be easily filtered out if someone wants to analyze the syslog and doesn't want to see them. So, there should be no harm in some printouts that go out every few seconds.

amnonh commented:
> > Also, I think spamming the logs with this is wrong. It should be made available over a cql virtual table, and then sampled/displayed/logged via the monitoring stack.
> 

My view that it should be a system table not a virtual table, and I actually have a plan to create such code once the work on this PR will be done.


> @amnonh is there a way to push the data from the virtual table into Prometheus somehow?
> 

Theoretically yes, in practice, no, the problem with all the top-k, is high churn rate, this is something Prometheus and Victoria metrics (which is doing a slightly better job) are bad at.

What we can do is have a different database that reads from the virtual table and store it in its own table, but I argue, that it should be Scylla itself, so instead of having virtual table, store it in a system table, with time-window compaction and configurable TTL.

avikivity commented:
It makes sense, virtual table for snapshot data, but here we want historical data, so a system table is indicated.

avikivity commented:
> > If the top-k configuration is wrong, you'll get incorrect data. In some cases you'll see the hot partition even with incorrect configuration, but in other cases it can be lost.
> 
> Could you, please, clarify what do you refer by "incorrect configuration" here?


Incorrect combination of duration/capacity.

> 
> > Also, I think spamming the logs with this is wrong. It should be made available over a cql virtual table, and then sampled/displayed/logged via the monitoring stack.
> 
> Using syslog + Locki allows us a very fast integration with the existing Scylla Monitoring stack (this is what you see on a screenshot).
> 
> It's unclear to me how to store historical data originally read from the virtual table. And the whole point of a Top-K graph is to have a historical data and not only the current state like we have for a `system.clients` table today.

@amnonh suggested a real table, and I agree.

> 
> @amnonh is there a way to push the data from the virtual table into Prometheus somehow?
> 
> Also, @avikivity, note that these new printouts go in a separate logger and hence can be easily filtered out if someone wants to analyze the syslog and doesn't want to see them. So, there should be no harm in some printouts that go out every few seconds.

Logs are accessible to the database operators, but often not to the application operators.

vladzcloudius commented:
> > > If the top-k configuration is wrong, you'll get incorrect data. In some cases you'll see the hot partition even with incorrect configuration, but in other cases it can be lost.
> > 
> > 
> > Could you, please, clarify what do you refer by "incorrect configuration" here?
> 
> Incorrect combination of duration/capacity.

@avikivity this feature is like a Probabilistic Tracing. You need to know how to use it in order to get a meaningful result.

The general idea is to have it always enabled and this PR allows you trade the overhead for precision by specifying a portion of queries that will accounted - just like a Probabilistic Tracing or `perf` does. 
The moment you realize that you see that there is no "incorrect configuration" - only incorrect expectations.

> 
> > > Also, I think spamming the logs with this is wrong. It should be made available over a cql virtual table, and then sampled/displayed/logged via the monitoring stack.
> > 
> > 
> > Using syslog + Locki allows us a very fast integration with the existing Scylla Monitoring stack (this is what you see on a screenshot).
> > It's unclear to me how to store historical data originally read from the virtual table. And the whole point of a Top-K graph is to have a historical data and not only the current state like we have for a `system.clients` table today.
> 
> @amnonh suggested a real table, and I agree.

I think that table, just like for Tracing today, is going to make us more trouble than good. We should not store the tracing information in the DB we are tracing - this creates an unnecessary noise in a form of compactions for example.

I personally don't see any advantage of storing this information in any form of table compared to dumping it in the sys log. There are some disadvantages I mentioned above however.

Why can't we continue this discussion and, if decided that we **really want** this info in some table as a follow up? This PR keeps things very simple and allows us to have a so long waited solution for our users/customers.

> 
> > @amnonh is there a way to push the data from the virtual table into Prometheus somehow?
> > Also, @avikivity, note that these new printouts go in a separate logger and hence can be easily filtered out if someone wants to analyze the syslog and doesn't want to see them. So, there should be no harm in some printouts that go out every few seconds.
> 
> Logs are accessible to the database operators, but often not to the application operators.

Not sure I understand your point here. Why would this statement be relevant to this discussion. Nobody is supposed to read these log lines - they will be processed by syslog processing SW like Loki and presented to the user as a graph like you can see on the screenshot.

vladzcloudius commented:
> What we can do is have a different database that reads from the virtual table and store it in its own table, but I argue, that it should be Scylla itself, so instead of having virtual table, store it in a system table, with time-window compaction and configurable TTL.

@amnonh we need this info to be presented on the Grafana graph. Does Grafana knows to read data from a CQL source? How all this is going to work? 
Again, @avikivity and @amnonh, please, keep in mind that the ultimate goal is to have this information to be **constantly** collected so that one can see what was going on in the past **on the graph** - not in some table! 

Loki allows us to do so **today** - hence we dump this info into the syslog.
Can we have the same with the table, Amnon? If "yes", please, could you describe the full solution with the table that will allow having such a graph?

amnonh commented:
> @amnonh we need this info to be presented on the Grafana graph. Does Grafana knows to read data from a CQL source? How all this is going to work? Again, @avikivity and @amnonh, please, keep in mind that the ultimate goal is to have this information to be **constantly** collected so that one can see what was going on in the past **on the graph** - not in CQL table.
> 

Few things, first I'm not against log as an output. It has the benefit of running out side of the monitored system and it's easy to use.
Second, my main point that if you use CQL (which I like) it should be a table not a virtual table.
Third, there is a grafana CQL plugin, we have out own plugin for a couple of years now (I just released a new version last week) and there's an official plugin, which is official on one hand but lack some capabilities on the other.

> Working with a complex tracing info in the table is hard. We want something that is simple. Graph - is simple.
> 
> Loki allows us to do so **today** - hence we dump this info into the syslog. Can we have the same with the table, Amnon? If "yes", please, could you describe the full solution with the table that will allow having such a graph?

Again, I like Loki, I think using logger is a good idea and we should start with that, but you can creat a graph from a table output as well.

avikivity commented:
Is there a way to set the top-k capacity and period?

igorribeiroduarte commented:
> Is there a way to set the top-k capacity and period?

@avikivity, yes. The following configs are added for that in this PR:
`persistent_toppartitions_publish_interval_sec` and `persistent_toppartitions_capacity`

vladzcloudius commented:
> > @amnonh we need this info to be presented on the Grafana graph. Does Grafana knows to read data from a CQL source? How all this is going to work? Again, @avikivity and @amnonh, please, keep in mind that the ultimate goal is to have this information to be **constantly** collected so that one can see what was going on in the past **on the graph** - not in CQL table.
> 
> Few things, first I'm not against log as an output. It has the benefit of running out side of the monitored system and it's easy to use. Second, my main point that if you use CQL (which I like) it should be a table not a virtual table. Third, there is a grafana CQL plugin, we have out own plugin for a couple of years now (I just released a new version last week) and there's an official plugin, which is official on one hand but lack some capabilities on the other.
> 
> > Working with a complex tracing info in the table is hard. We want something that is simple. Graph - is simple.
> > Loki allows us to do so **today** - hence we dump this info into the syslog. Can we have the same with the table, Amnon? If "yes", please, could you describe the full solution with the table that will allow having such a graph?
> 
> Again, I like Loki, I think using logger is a good idea and we should start with that, but you can creat a graph from a table output as well.

Could you, please, clarify how exactly? I did try to do that but couldn't find how to implement `rate()` and `sum()` aggregates on top of that.

amnonh commented:
> 
> Could you, please, clarify how exactly? I did try to do that but couldn't find how to implement `rate()` and `sum()` aggregates on top of that.

The basic query only supports what Scylla supports, but you can do a bit more, if you check the transform part of the panels

vladzcloudius commented:
> > Could you, please, clarify how exactly? I did try to do that but couldn't find how to implement `rate()` and `sum()` aggregates on top of that.
> 
> The basic query only supports what Scylla supports, but you can do a bit more, if you check the transform part of the panels

I checked and wasn't able to achieve what I described above.
If you know how - please, share.

igorribeiroduarte commented:
> > Is there a way to set the top-k capacity and period?
> 
> @avikivity, yes. The following configs are added for that in this PR: `persistent_toppartitions_publish_interval_sec` and `persistent_toppartitions_capacity`

ping @avikivity

igorribeiroduarte commented:
> > > Is there a way to set the top-k capacity and period?
> > 
> > 
> > @avikivity, yes. The following configs are added for that in this PR: `persistent_toppartitions_publish_interval_sec` and `persistent_toppartitions_capacity`
> 
> ping @avikivity

ping2 @avikivity

epaladugu commented:
Nice PR! Looking forward for this Top-K graph feature to become available! We, at Tubi, would really appreciate this feature.

igorribeiroduarte commented:
Following the call with @avikivity  we agreed to double check what were the blockers to use CQL as a data source for Top-K data.
We will perform additional investigation and will update once we have the results.
Stay tuned.
